src/test/typescript/typescript-analyzer_test.ts

```diff
@@ -0,0 +1,105 @@
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {assert} from 'chai';
+import * as ts from 'typescript';
+
+import {AnalysisContext} from '../../core/analysis-context';
+import {TypeScriptAnalyzer} from '../../typescript/typescript-analyzer';
+import {UrlLoader} from '../../url-loader/url-loader';
+import {UrlResolver} from '../../url-loader/url-resolver';
+
+class TestUrlResolver implements UrlResolver {
+  canResolve(url: string) {
+    return url === '/typescript/test.ts';
+  }
+
+  resolve(url: string) {
+    return url;
+  }
+}
+
+class TestUrlLoader implements UrlLoader {
+  canLoad(url: string) {
+    return url === '/typescript/test.ts';
+  }
+
+  load(url: string): Promise<string> {
+    if (url === '/typescript/test.ts') {
+      return Promise.resolve(`
+class A extends HTMLElement {
+  foo() { return 'bar'; }
+}
+`);
+    } else {
+      throw new Error(`cannot load file ${url}`);
+    }
+  }
+}
+
+suite('TypeScriptParser', () => {
+  let typescriptAnalyzer: TypeScriptAnalyzer;
+  let analysisContext: AnalysisContext;
+
+  setup(() => {
+    const urlLoader = new TestUrlLoader();
+    const urlResolver = new TestUrlResolver();
+    analysisContext = new AnalysisContext({urlLoader, urlResolver});
+    typescriptAnalyzer = new TypeScriptAnalyzer(analysisContext);
+  });
+
+  suite('parse()', () => {
+
+    test('parses classes', async() => {
+      const fileName = '/typescript/test.ts';
+      // This puts a document into the scanned document cache
+      await analysisContext.scan(fileName);
+      const program = typescriptAnalyzer.analyze(fileName);
+      const checker = program.getTypeChecker();
+
+      assert.deepEqual(program.getRootFileNames(), [fileName]);
+
+      // Get the HTMLElement type from the DOM module
+      let htmlElement: ts.Type;
+      const domSource = program.getSourceFile('/$lib/DOM.d.ts');
+      ts.forEachChild(domSource, (node) => {
+        if (node.kind === ts.SyntaxKind.InterfaceDeclaration) {
+          const innerface = node as ts.InterfaceDeclaration;
+          if (innerface.name.getText() === 'HTMLElement') {
+            htmlElement = checker.getTypeAtLocation(innerface);
+          }
+        }
+      });
+
+      // Get class A and assert that it extends HTMLElement
+      const sourceFile = program.getSourceFile(fileName);
+      ts.forEachChild(sourceFile, (node) => {
+        if (node.kind === ts.SyntaxKind.ClassDeclaration) {
+          const clazz = node as ts.ClassDeclaration;
+          if (clazz.name && clazz.name.getText() === 'A') {
+            const type = checker.getTypeAtLocation(clazz);
+            const baseTypes = checker.getBaseTypes(type as ts.InterfaceType);
```

Comment:
Well there's the cast to ClassDecoration that immediately follows a comparison for same, which is just a discrepancy in the type checker (can't do type discrimination on enum instances).

The type type should be able to do the same (even if we might not bother with the assertion here).